File: engines/pink/objects/cpage.cpp
#include "cpage.h"
#include "carchive.h"
#include "cruntimeclass.h"

#include "common/str.h"
#include "common/array.h"

#include "cactor.h"

namespace Pink {

	CPage::CPage() : directory(NULL), actors(NULL), lead_actor(NULL) {
	}

	CPage::~CPage() {
		delete directory;
		delete actors;
	}

	void CPage::deserialize(CArchive &archive) {
		this->CNamedObject::deserialize(archive);
		directory = archive.readCString();
		actors = (Common::Array<CActor *> *)archive.readCObArray(CActor::RuntimeClass());
		for (uint32 i = 0; i < actors->size(); i++) {
			CActor *actor = actors->at(i);
			_actors_map[*(actor->name)] = actor;
		}
	}

	void CPage::readFromOrb(CArchive &archive) {
		archive.mapObject(this);
		this->CPage::deserialize(archive);
	}

	void CPage::init(PinkEngine *pink) {
		for (uint i = 0; i < actors->size(); i++) {
			actors->at(i)->init(pink);
		}
	}

	void CPage::execute(PinkEngine *pink) {
		for (uint i = 0; i < actors->size(); i++) {
			actors->at(i)->execute(pink);
		}
	}

	void CPage::destroy(PinkEngine *pink) {
		for (uint i = 0; i < actors->size(); i++) {
			actors->at(i)->destroy(pink);
		}
	}

	CActor *CPage::getActor(const Common::String &name) {
		return _actors_map.getVal(name, NULL);
	}
};

File: engines/pink/OrbFile/CPage.py
from CArchive import *

from CObject import *
from Environment import *
from CSequence import *
from CActor import *
from CHandler import *
from CSideEffect import *

class CPage(CNamedObject):
    def __init__(self):
        CNamedObject.__init__(self)
        self.directory = ''
        self.actors = []
        self.leadActor = None

    def readFromOrb(self, archive):
        archive.mapObject(self)
        CNamedObject.deserialize(self, archive)
        self.directory = archive.readCString()
        self.actors = archive.readCObArray()
        for actor in self.actors:
            if isinstance(actor, CLeadActor):
                self.leadActor = actor
                break
        
    def _exportAttributes(self):
        text = CNamedObject._exportAttributes(self)
        text += ["Directory: %s" % self.directory]
        return text

    def _exportSubobjects(self):
        text = CNamedObject._exportSubobjects(self)
        for actor in self.actors:
            text += actor.exportText()
        return text
        

CArchive.register(CPage)

class CGamePage(CPage):
    def __init__(self):
        CPage.__init__(self)
        self.cursorMgr = CCursorMgr()
        self.walkMgr = CWalkMgr()
        self.sequencer = CSequencer()
        self.handlers = []

    def readFromOrb(self, archive):
        archive.mapObject(self.cursorMgr)
        archive.mapObject(self.walkMgr)
        archive.mapObject(self.sequencer)
        
        CPage.readFromOrb(self, archive)

        self.leadActor = archive.readObject(CLeadActor)
        self.cursorMgr.deserialize(archive)
        self.walkMgr.deserialize(archive)
        self.sequencer.deserialize(archive)
        self.handlers = archive.readCObArray()

    def deserialize(self, archive):
        CNamedObject.deserialize(self, archive)
        self.module = archive.readObject()

    def _exportSubobjects(self):
        text = CPage._exportSubobjects(self)
        for handler in self.handlers:
            text += handler.exportText()
        text += self.walkMgr.exportText()
        text += self.sequencer.exportText()
        return text

CArchive.register(CGamePage)

__all__ = ["CPage", "CGamePage"]
